src/java/main/org/apache/zookeeper/ClientCnxn.java

```diff
@@ -1495,10 +1504,21 @@ public ReplyHeader submitRequest(RequestHeader h, Record request,
         Packet packet = queuePacket(h, r, request, response, null, null, null,
                 null, watchRegistration, watchDeregistration);
         synchronized (packet) {
+            long waitStartTime = System.currentTimeMillis();
             while (!packet.finished) {
-                packet.wait();
+                packet.wait(requestTimeout);
```

Comment:
This unconditionally adds a timeout is a pretty significant semantic change. Whether or not we should wait indefinitely is a separate issue to discuss later, but from a compatibility point of view we'd at least provide a configuration option to let user opt in this feature rather than having this turned on by default. So by default there is no timeout and client still waits indefinitely, user who wants to opt in to turn on the timeout needs to explicit say so and provide a value they think make sense to them. Otherwise, we'd have to solve the problem of what the default timeout value we should set in ZK config if this feature is turned on by default, and that problem is hard and very likely there is no default value that makes everyone happy.

Reply:
disabled this feature by default. any non positive value will disable this feature